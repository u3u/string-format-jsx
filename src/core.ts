export interface FormatElementsOptions<T> {
  components?: Record<string, (children: T | string, key: number) => T>;
  createElement?(type: string, children: T | string, key: number): T;
}

export const formatElements = <T>(text: string, options: FormatElementsOptions<T> = {}) => {
  if (!text) return text;

  const { components, createElement } = options;
  const parts = text.replaceAll(nlRe, '').split(tagParsingRegex);

  if (parts.length === 1) return text;

  const tree: (T | string)[] = [];
  const before = parts.shift();

  if (before) tree.push(before);

  const elements = getElements(parts);

  for (const [key, [type, children, after]] of elements.entries()) {
    const render = components?.[type];
    const child = children && (formatElements<T>(children, options) as T | string);

    tree.push(
      render
        ? render?.(child, key)
        : //
          createElement
          ? createElement?.(type, child, key)
          : child,
    );

    if (after) tree.push(after);
  }

  return tree;
};

export const getElements = (parts: string[]): string[][] => {
  if (!parts.length) return [];

  const [paired, children, unpaired, after] = parts.slice(0, 4);

  return [
    //
    [paired || unpaired, children, after],
    ...getElements(parts.slice(4, parts.length)),
  ];
};

export const nlRe = /\r\n|\r|\n/gu;

export const tagParsingRegex = /<(\w+) *>(.*?)<\/\1 *>|<(\w+) *\/>/u;
