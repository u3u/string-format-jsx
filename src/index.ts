export * from './core';
export * from './interpolation';
