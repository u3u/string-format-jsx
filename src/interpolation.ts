export interface InterpolationOptions {
  /**
   * Custom interpolation regexp
   *
   * @default /\{(.*?)\}/gu
   */
  regex?: RegExp;
  values?: Record<string, number | string>;
}

/**
 * @example
 *   const result = interpolation('{name} 你好，我今年 {age} 岁，我的爱好是 {hobby} 。', {
 *     values: {
 *       age: 18,
 *       hobby: '唱，跳，rap，篮球',
 *       name: '蔡徐坤',
 *     },
 *   });
 */
export const interpolation = (text: string, options: InterpolationOptions = {}) => {
  const { regex = /\{(.*?)\}/gu, values } = options;

  let match: RegExpExecArray | null;
  let result = text;

  while ((match = regex.exec(text))) {
    const [name, key] = match;
    const val = values?.[key];

    if (val !== undefined) result = result.replaceAll(name, String(val));
  }

  return result;
};
