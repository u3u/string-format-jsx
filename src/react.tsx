import { createElement, type ReactNode } from 'react';
import { formatElements as format, type FormatElementsOptions as Options } from './core';

export interface FormatElementsOptions extends Options<ReactNode> {}

/**
 * @example
 *   const result = formatElements(
 *     dedent`
 *       Hello, <component>The number is <b>{count}</b></component>
 *       <br />
 *       The number is <b>{count}</b>.
 *     `,
 *     {
 *       components: {
 *         component: (children) => <div>{children}</div>,
 *       },
 *     },
 *   );
 */
export const formatElements = (text: string, options: FormatElementsOptions = {}) => {
  return format<ReactNode>(text, {
    createElement: (type, children, key) => createElement(type, { key }, children),
    ...options,
  });
};
