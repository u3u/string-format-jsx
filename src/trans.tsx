import React from 'react';
import { interpolation, type InterpolationOptions } from './interpolation';
import { formatElements, type FormatElementsOptions } from './react';

export interface TransProps extends InterpolationOptions {
  components?: FormatElementsOptions['components'];
  text: string;
}

export const Trans = (props: TransProps) => {
  const { components, regex, text, values } = props;

  return (
    <>
      {formatElements(
        interpolation(text, {
          regex,
          values,
        }),
        { components },
      )}
    </>
  );
};
