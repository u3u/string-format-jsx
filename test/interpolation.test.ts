import { expect, it } from 'vitest';
import { interpolation } from '../src';

it('should match snapshots', () => {
  const result = [
    interpolation('Hello, {name} ({count})', {
      values: {
        count: 0,
        name: 'u3u',
      },
    }),

    interpolation('Hello, {name} ({count})', {
      values: {
        name: 'u3u',
      },
    }),

    interpolation('Hello, {{name}}.', {
      regex: /\{\{(.*?)\}\}/gu,

      values: {
        name: 'qwq',
      },
    }),

    interpolation('{name} 你好，我今年 {age} 岁，我的爱好是 {hobby} 。', {
      values: {
        age: 18,
        hobby: '唱，跳，rap，篮球',
        name: '蔡徐坤',
      },
    }),
  ];

  expect(result).toMatchSnapshot();
});
