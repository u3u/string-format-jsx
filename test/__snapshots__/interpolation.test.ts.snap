// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshots 1`] = `
[
  "Hello, u3u (0)",
  "Hello, u3u ({count})",
  "Hello, qwq.",
  "蔡徐坤 你好，我今年 18 岁，我的爱好是 唱，跳，rap，篮球 。",
]
`;
